File: 백준/Gold/17396. 백도어/백도어.py
from heapq import heappush, heappop
from dataclasses import dataclass

@dataclass
class Edge:
    id: int
    cost : int

    def __lt__(self, other):
        return self.cost < other.cost


if __name__ == "__main__":
    N, M = map(int,input().split())
    paths = list(map(int,input().split()))
    INF = float('inf')
    dist = [INF for _ in range(N)]

    adj = [[] for _ in range(N)]

    for _ in range(M):
        u, v, cost = map(int,input().split())
        adj[u].append(Edge(v, cost))
        adj[v].append(Edge(u, cost))


    dist[0] = 0
    pq = []
    heappush(pq, Edge(0, dist[0]))

    while pq:
        cur : Edge = heappop(pq)

        if dist[cur.id] < cur.cost:
            continue

        for nxt in adj[cur.id]:
            if paths[nxt.id] == 1 and nxt.id != (N-1):
                continue

            if dist[nxt.id] > dist[cur.id] + nxt.cost:
                dist[nxt.id] = dist[cur.id] + nxt.cost
                heappush(pq, Edge(nxt.id, dist[nxt.id]))

    print(dist[N-1] if dist[N-1] != INF else -1)

File: 백준/Gold/17396. 백도어/백도어.java
import java.util.*;
import java.io.*;

public class Main{
	public static void main(String[] args) throws IOException{
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		st = new StringTokenizer(br.readLine());
		N = Integer.parseInt(st.nextToken());
		M = Integer.parseInt(st.nextToken());
		paths = new int[N];
		dist = new long[N];
		adj = new List[N];
		st = new StringTokenizer(br.readLine());
		
		for(int i = 0; i < N; i++) {
			paths[i] = Integer.parseInt(st.nextToken());
			adj[i] = new ArrayList<>();
			dist[i] = INF;
		}
		
		for(int i = 0; i < M; i++) {
			st = new StringTokenizer(br.readLine());
			int from = Integer.parseInt(st.nextToken());
			int to = Integer.parseInt(st.nextToken());
			int cost = Integer.parseInt(st.nextToken());
			adj[from].add(new Edge(to, cost));
			adj[to].add(new Edge(from, cost));
			
		}
		
		dist[0] = 0;
		PriorityQueue<Edge> pq = new PriorityQueue<>();
		pq.add(new Edge(0, dist[0]));
		
		
		while(!pq.isEmpty()) {
			Edge cur = pq.poll();
			
			if(dist[cur.id] < cur.cost)
				continue;
			
			for(Edge nxt : adj[cur.id]) {
				if(paths[nxt.id] == 1 && nxt.id != (N-1))
					continue;
				
				if(dist[nxt.id] > dist[cur.id] + nxt.cost) {
					dist[nxt.id] = dist[cur.id] + nxt.cost;
					pq.add(new Edge(nxt.id, dist[nxt.id]));
				}
			}
		}
		
		System.out.println(dist[N-1] == INF ? -1 : dist[N-1]);
		
	}
	static long dist[];
	static long INF = Long.MAX_VALUE - 1;
	static int paths[];
	static StringTokenizer st;
	
	static int N,M;
	static List<Edge> adj[];
	
	static class Edge implements Comparable<Edge>{
		int id;
		long cost;
		public Edge(int id, long cost) {
			this.id = id;
			this.cost = cost;
		}
		
		public int compareTo(Edge o) {
			return Long.compare(this.cost, o.cost);
		}
	}
}

